tests/tcpclientsrc.rs

```diff
@@ -0,0 +1,152 @@
+// Library General Public License for more details.
+//
+// You should have received a copy of the GNU Library General Public
+// License along with this library; if not, write to the
+// Free Software Foundation, Inc., 51 Franklin Street, Suite 500,
+// Boston, MA 02110-1335, USA.
+
+extern crate glib;
+use glib::prelude::*;
+
+extern crate gstreamer as gst;
+use gst::prelude::*;
+
+use std::io::Write;
+use std::sync::{Arc, Mutex};
+use std::{thread, time};
+
+fn init() {
+    use std::sync::{Once, ONCE_INIT};
+    static INIT: Once = ONCE_INIT;
+
+    INIT.call_once(|| {
+        gst::init().unwrap();
+
+        #[cfg(debug_assertions)]
+        {
+            use std::path::Path;
+
+            let mut path = Path::new("target/debug");
+            if !path.exists() {
+                path = Path::new("../target/debug");
+            }
+
+            gst::Registry::get().scan_path(path);
+        }
+        #[cfg(not(debug_assertions))]
+        {
+            use std::path::Path;
+
+            let mut path = Path::new("target/release");
+            if !path.exists() {
+                path = Path::new("../target/release");
+            }
+
+            gst::Registry::get().scan_path(path);
+        }
+    });
+}
+
+#[test]
+fn test_push() {
+    init();
+
+    let handler = thread::spawn(move || {
+        use std::net;
+
+        let listener = net::TcpListener::bind("0.0.0.0:5000").unwrap();
+        for stream in listener.incoming() {
+            let buffer = [0; 160];
+            let mut socket = stream.unwrap();
+            for _ in 0..3 {
+                let _ = socket.write(&buffer);
+                thread::sleep(time::Duration::from_millis(20));
+            }
+            break;
+        }
+    });
+
+    let pipeline = gst::Pipeline::new(None);
+
+    let tcpclientsrc = gst::ElementFactory::make("ts-tcpclientsrc", None).unwrap();
+    let appsink = gst::ElementFactory::make("appsink", None).unwrap();
+    appsink.set_property("sync", &false).unwrap();
+    appsink.set_property("async", &false).unwrap();
+
+    pipeline.add_many(&[&tcpclientsrc, &appsink]).unwrap();
+    tcpclientsrc.link(&appsink).unwrap();
+
+    let caps = gst::Caps::new_simple("foo/bar", &[]);
+    tcpclientsrc.set_property("caps", &caps).unwrap();
+    tcpclientsrc.set_property("port", &(5000u32)).unwrap();
+
+    appsink.set_property("emit-signals", &true).unwrap();
+
+    let samples = Arc::new(Mutex::new(Vec::new()));
+
+    let samples_clone = samples.clone();
+    appsink
+        .connect("new-sample", true, move |args| {
+            let appsink = args[0].get::<gst::Element>().unwrap();
+
+            let sample = appsink
+                .emit("pull-sample", &[])
+                .unwrap()
+                .unwrap()
+                .get::<gst::Sample>()
+                .unwrap();
+
+            let mut samples = samples_clone.lock().unwrap();
+
+            samples.push(sample);
+            if samples.len() == 3 {
```

Comment:
This is TCP, a stream. So it might be that we get all `3*160` bytes in one buffer, in 3 buffers, in 100 buffers. Nobody knows :) You should count the bytes and not the number of buffers here

Reply:
Fixed it 8b1dd1a